### ISSUE TITLE
Not able to Update LookUp property

### ISSUE THREAD

**Issue by gopaldahal** [pnp/pnpcore#502]
## Category

- [x] Bug

## Describe the bug

Using CSOM, It is possible to update the lookup property with a string value, which does not particularly require its lookupvalue property type.

For example
item["MetaInfo"] = "some XML values"
It is possible to read back the same string value during the read.
I need a way to update value in a same way on PnP.

## Steps to reproduce
I tried the same on PnP, However, am not able to set the property. It neither throws an exception.
```
var parentList = await context.Web.Lists.GetByTitleAsync("AddTest", l => l.EnableModeration, l => l.TemplateType, l => l.EnableVersioning, l => l.Title, l => l.Fields, l => l.ContentTypes);
            var parentWeb = await context.Web.GetAsync(s => s.Language, s=>s.Url);
            var parentFields = parentList.Fields.ToList();

            List<Dictionary<string, object>> propertiesToUpdate = new List<Dictionary<string, object>>();
            var prop = new Dictionary<string, object>();
            prop.Add("Title", "My title 1");
            prop.Add("MetaInfo", "abc");
            propertiesToUpdate.Add(prop);
            var prop2 = new Dictionary<string, object>();
            prop2.Add("Title", "My title 2");
            propertiesToUpdate.Add(prop2);

            var uploadedListItems = new List<IListItem>();
            foreach (var propItem in propertiesToUpdate)
            {     
                try
                {
                    var item = await parentList.Items.AddBatchAsync(propItem);
                    uploadedListItems.Add(item);
                }
                catch(Exception ex)
                {
                }
            }
            var batchResponses = await context.ExecuteAsync(false);

            await uploadedListItems.First().LoadAsync(d => d.All);

            foreach (var batchResponse in batchResponses)
            {

            }
```

## Expected behavior
It should have been updated with no issue
## Environment details (*development & target environment*)

- **SDK version**: [1.1.0 ]
- **OS**: [Windows 10]
- **SDK used in**: [ Console App ]
- **Framework**: [.NET Framework v4.7.2]
- **Tooling**: [Visual Studio 2019]

### Additional context

*Thanks for your contribution! Sharing is caring.*

**Comment by jansenbe** [comment]
Related to #485, which I'm working on

**Comment by jansenbe** [comment]
The accidental "clearing" of special fields (like lookup or user) has been fixed via #485. Reading your question again and trying to understand the ask here. You set a lookup field by setting the lookupid...this will then make SharePoint display the relevant values in the columns that are taken from the looked-up list item. Using the lookupid property is possible today (see https://pnp.github.io/pnpcore/using-the-sdk/listitems-fields.html#lookup-fields), so not sure what you mean with "update the lookup property with a string value, which does not particularly require its lookupvalue property type"...can you provide more context @gopaldahal ?

**Comment by gopaldahal** [comment]
@jansenbe, Using CSOM I need not use create NewFieldLookupValue type object, Which Means I used to assign the string value of Lookvalue of Id (In string form).
I got with the below implementation.
```
// Add a lookup field to an item in the connected lookup list with id 4
values.Add("MyField", myField.NewFieldLookupValue(4));
```
I am preparing with an example set. will update soon on the same thread. Till then please do not close the thread. Please :)

**Comment by jansenbe** [comment]
@gopaldahal : I'm not closing the issue, no worries :-)

The thing however is that since we use the `AddValidateUpdateItemUsingPath` REST call for adding and `ValidateUpdateListItem` REST call for updating we need information about the type of field being added. That's the reason we recommend ensuring the fields of a list are loaded before you started adding/updating items to that list. You can combined loading the fields of a list with loading the list itself, so the performance overhead is minimal. See https://pnp.github.io/pnpcore/using-the-sdk/listitems-intro.html#reading-list-items for more details.

**Comment by gopaldahal** [comment]
Hi @jansenbe,

As I mentioned earlier, using CSOM I am able to insert the XML string(or any string) on the "MetaInfo" (Look up type) property (The property is the default property of SharePoint itself which holds metadata of item). 
I am doing this to keep additional information on items for my additional business purpose, so that I can track/read when needed.

Using CSOM what I am currently doing is,
```
item["MetaInfo"] = "{DFC8691F-2432-4741-B780-3CAE3235A612}:SW|<?xml version="1.0" encoding="utf-16"?><myxml xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"></myxml >"
item.Update();
```

The same is can be read by making a CSOM call.

If I make the read of the same property using PnP,
The property value of "MetaInfo" is appeared to be as

item["MetaInfo"] = 73;#vti_encoding:SR|utf8-nl
{DFC8691F-2432-4741-B780-3CAE3235A612}:SW|<?xml version="1.0" encoding="utf-16"?><myxml xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"></myxml >

Thus I am in search of adding the same sort of data using PnP itself.
But the current implementation just support the lookup id as

```
IField myField = await myList.Fields.AddLookupAsync("MyField", new FieldLookupOptions()
{
    Group = "Custom Fields",
    AddToDefaultView = true,
    LookupListId = sitePages.Id,
    LookupFieldName = "Title",
});

// Create a list item key/value pair collection
Dictionary<string, object> values = new Dictionary<string, object>()
{
    { "Title", "Item1" },
};

// Add a lookup field to an item in the connected lookup list with id 4
values.Add("MyField", myField.NewFieldLookupValue(4));

// Persist the item
var addedItem = await myList.Items.AddAsync(values);

// Update the lookup field
(addedItem["MyField"] as IFieldLookupValue).LookupId = 8;
// Or clear the lookup field
addedItem["MyField"] = null;

// Update the item on the server
await addedItem.UpdateAsync();
```
which just supports the lookup value in integer id form

I hope I am making clear the scenario.
Please let me know if I need to add more.

Thank you

**Comment by jansenbe** [comment]
Hi @gopaldahal , have been doing some checking on this and to me updating MetaInfo by appending data to it feels like a hack. Since you've been doing this, it works, but the correct way to add custom data to a list item is by adding the needed (hidden) columns for your custom data. As there's server-side code that parses this field adding arbitrary content to it is not recommended and to be honest, I've never seen this use case.

Reading the metainfo should be possible I think (but did not test): did you try to load  the `FieldValuesAsText` list item property and then check for the MetaInfo field (` listItem.FieldValuesAsText["MetaInfo"]`)?

**Comment by gopaldahal** [comment]
Hi @jansenbe 
![image](https://user-images.githubusercontent.com/27617624/124371916-221deb00-dca6-11eb-844e-eb6523722b8e.png)
I did read with ### FieldValuesAsText but working as expected.

**Comment by jansenbe** [comment]
@gopaldahal : are you good with closing this issue?

**Comment by gopaldahal** [comment]
Hi @jansenbe,
I still didn't get the solution for my problem.
I am working for years on setting and retrieving "MetaInfo" Information using CSOM with explained set of data.
And SharePoint itself supports from its end.

This is the case of mandatory to be supported on Pnp from my end.

Since SharePoint itself support the case, What I suppose is we need to address the scenario on PnP too.
There must be some way to achieve it.
Thus request you to let it remain open until we get it resolved.

Thank you

Thankyou for the contribution! Sharing is caring!

**Comment by jansenbe** [comment]
@gopaldahal : can you provide a fiddler trace showing how this is set using CSOM and send it to bjansen@microsoft.com. When you do a SystemUpdate() or UpdateOverwriteVersion() in PnP Core we under the covers fall back to CSOM...I want to cross check if this approach can unblock this scenario for you.

**Comment by gopaldahal** [comment]
> @gopaldahal : can you provide a fiddler trace showing how this is set using CSOM and send it to [bjansen@microsoft.com](mailto:bjansen@microsoft.com). When you do a SystemUpdate() or UpdateOverwriteVersion() in PnP Core we under the covers fall back to CSOM...I want to cross check if this approach can unblock this scenario for you.

Sure. I will come to this soon.

**Comment by gopaldahal** [comment]
Hi @jansenbe,

I have sent you an email with the attachments that hold the request-response of the request made through CSOM.
```
        public void UpdateItemProperties()
        {
            var web = ClientContext.Web;
            //var oList = ClientContext.Web.Lists.GetById(new Guid("31e38d69-ee68-4eea-8cc4-0f7254cd1a72"));
            var oList = ClientContext.Web.Lists.GetByTitle("HeyramList");
            ListItem listItem = oList.GetItemById(75);
            ClientContext.Load(listItem);
            ClientContext.Load(oList, d => d.Fields);
            ClientContext.ExecuteQuery();

            var metaInfoField = oList.Fields.First(d => d.InternalName == "MetaInfo");
            var fieldTypeKind = metaInfoField.FieldTypeKind; //lookup
            listItem["MetaInfo"] = "{DFC8691F-2432-4741-B780-3CAE3235A612}:SW|MyStringWithXmlValues";
            listItem["Title"] = "jayshreeram4";
            listItem.Update();
            ClientContext.ExecuteQuery();
}
```
This is the working code I am using to update the "MetaInfo" (Lookup) property using CSOM.

Please let me know if I need to add more.

Thank you

Sharing is Caring!!

**Comment by jansenbe** [comment]
Hey @gopaldahal ,

In my testing your scenario (updating the MetaInfo field) works when you use `SystemUpdate` or `SystemUpdateOverwriteVersion`. These two methods are implemented using CSOM under the covers (as there's no REST equivalent). Below is my test code:

```csharp
[TestMethod]
public async Task SystemUpdateMetaInfo()
{
    TestCommon.Instance.Mocking = false;
    string listTitle = "SystemUpdateMetaInfo";

    try
    {
        using (var context = await TestCommon.Instance.GetContextAsync(TestCommon.TestSite))
        {
            // Create a new list
            var myList = context.Web.Lists.FirstOrDefault(p => p.Title == listTitle);

            if (TestCommon.Instance.Mocking && myList != null)
            {
                Assert.Inconclusive("Test data set should be setup to not have the list available.");
            }

            if (myList == null)
            {
                myList = await context.Web.Lists.AddAsync(listTitle, ListTemplateType.GenericList);
                // Enable versioning
                myList.EnableVersioning = true;
                await myList.UpdateAsync();
            }

            // Add items to the list
            for (int i = 0; i < 10; i++)
            {
                Dictionary<string, object> values = new Dictionary<string, object>
                {
                    { "Title", $"Item {i}" }
                };

                await myList.Items.AddBatchAsync(values);
            }
            await context.ExecuteAsync();

            // get first item and do a system update
            var first = myList.Items.AsRequested().First();

            first["MetaInfo"] = "{DFC8691F-2432-4741-B780-3CAE3235A612}:SW|MyStringWithXmlValues";
            first["Title"] = "okido";

            await first.SystemUpdateAsync();
        }

        using (var context2 = await TestCommon.Instance.GetContextAsync(TestCommon.TestSite, 1))
        {
            var myList2 = context2.Web.Lists.FirstOrDefault(p => p.Title == listTitle);
            await myList2.LoadAsync(p => p.Items.QueryProperties(p=>p.Title, p=>p.FieldValuesAsText));

            var first2 = myList2.Items.AsRequested().First();

            // verify the list item was updated and that we're still at version 1.0
            Assert.IsTrue(first2.Title == "okido");
            Assert.IsTrue(first2.FieldValuesAsText.Values["MetaInfo"].ToString().Contains("{DFC8691F-2432-4741-B780-3CAE3235A612}:SW|MyStringWithXmlValues"));
        }

    }
    finally
    {
        using (var contextFinal = await TestCommon.Instance.GetContextAsync(TestCommon.TestSite, 4))
        {
            var myList = contextFinal.Web.Lists.FirstOrDefault(p => p.Title == listTitle);

            // Cleanup the created list
            await myList.DeleteAsync();
        }
    }
}
```

**Comment by gopaldahal** [comment]
Hi @jansenbe,

Thank you for the update and sample code.
But my concern is regarding the Batch operations.
I have hundreds of items that need to be added to the list. Adding all the items on batch and updating the same items individually with system update does not justify me as a batch which ultimately increases the sever call.

My question is,
Do you mean, we can't achieve the "MetaInfo" Field to be added on the batch in any way? OR
We can archive this after immediate adding, and updating all items as batches?
Please help.

Sharing is Caring!!

**Comment by jansenbe** [comment]
@gopaldahal , let me check the add story...when using `SystemUpdate` or `UpdateOverwriteVersion` you can also apply batching, so you could have one batch that creates the items and one batch that updates them to set the "MetaInfo" field.

**Comment by jansenbe** [comment]
Using our REST based add methods it's not possible to set this field on list item add, but as mentioned above batching of the update requests does work. I've updated my test case (see ) or below snippet which adds/updates 10 items via 2 server roundtrips.

```csharp
// Add items to the list
for (int i = 0; i < 10; i++)
{
    Dictionary<string, object> values = new Dictionary<string, object>
    {
        { "Title", $"Item {i}" }
    };

    await myList.Items.AddBatchAsync(values);
}
await context.ExecuteAsync();

foreach(var item in myList.Items.AsRequested())
{
    item["MetaInfo"] = "{DFC8691F-2432-4741-B780-3CAE3235A612}:SW|MyStringWithXmlValues";
    item["Title"] = "okido";
    // Both work
    await item.SystemUpdateBatchAsync();
    //await item.UpdateOverwriteVersionBatchAsync();
}
await context.ExecuteAsync();
```

**Comment by jansenbe** [comment]
@gopaldahal : I don't see any other options for this edge case (there's very few people updating the metainfo field), so are you good with closing this issue?

**Comment by jansenbe** [comment]
@gopaldahal : are you in your current CSOM code adding a list item and setting the "MetaInfo" field in a single request?

**Comment by gopaldahal** [comment]
@jansenbe:

The interesting thing that I find on PnP.Core.SDK is its batch operation. The primary reason, I am testing on its behaviour is for the same reason.
There must be some workaround with this one and do not suggest you with its close.

Thank you

**Comment by jansenbe** [comment]
@gopaldahal : batching of `SystemUpdate` or `UpdateOverwriteVersion` works when updating the MetaInfo field of list items...what else do you need? Confused now...

**Comment by gopaldahal** [comment]
@jansenbe:
I am just giving concern on if there came up with other properties that cant go on batch during add.
Additionally, I am being confused about how can I differentiate the properties that can go on batch during **Add** and the properties that must be pushed with the **update**.
These are the reason, I don't want you to make this item close immediately.
If you are cent percent sure that property cannot be uploaded with BatchOnAdd, you can close the item.
Than you
Sharing is Caring!!

**Comment by jansenbe** [comment]
@gopaldahal : fields added by a user or application can be updated/added without issues, the fields you're having issues with are system fields. Some system fields can be set (like author/editor/modified/created if you update them via UpdateOverwriteVersion) but most can't and shouldn't. https://github.com/pnp/pnpcore/blob/dev/src/sdk/PnP.Core/Model/SharePoint/Core/Internal/BuiltInFields.cs is a class we internally use to define the "system" fields. I'm not convinced that stuffing information in MetaInfo in this manner is a supported scenario, it has always been working but that does not mean it's a supported scenario. If your application requires it to "hide" information into a list item without adding a new field then that's fine, but a better (and supported) approach would be to ListItem propertybag. This however will require more server calls, so less optimal for performance.

I'm going to close this issue now, but you know you can always create a new issue for a specific field issue once it comes up and we can discuss it.

**Comment by gopaldahal** [comment]
@jansenbe:
My apology but am neither convinced with your justification. I was just concerned with the behavior of SharePoint itself. If it supports batch, must have support for every field (not for some).  Since the property can be updated, it must have been supported with Add too.

I hope the demand of such behavior will come in bulk sooner. We may need to be prepare for the same though we have some secondary level of alternatives.

Thank you  
Sharing is Caring!!
